diesel/src/pg/expression/expression_methods.rs

```diff
@@ -810,6 +734,326 @@ pub trait PgTextExpressionMethods: Expression + Sized {
+    ///
+    /// This operator returns true whether an object is not a valid JSON ARRAY
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// # include!("../../doctest_setup.rs");
+    /// #
+    /// # fn main() {
+    /// #     run_test().unwrap();
+    /// # }
+    /// #
+    /// # fn run_test() -> QueryResult<()> {
+    /// #     use std::collections::Bound;
+    /// #     use diesel::sql_types::Text;
+    /// #
+    /// #     let conn = &mut establish_connection();
+    /// #
+    ///
+    /// let res = diesel::select(("123".into_sql::<Text>().is_not_json_array())).get_result::<bool>(conn)?;
+    /// assert_eq!(res, true);
+    /// let res = diesel::select(("abc".into_sql::<Text>().is_not_json_array())).get_result::<bool>(conn)?;
+    /// assert_eq!(res, true);
+    /// let res = diesel::select(("{\"products\": [1,2,3]}".into_sql::<Text>().is_not_json_array())).get_result::<bool>(conn)?;
+    /// assert_eq!(res, true);
+    /// let res = diesel::select(("[1,2,3]".into_sql::<Text>().is_not_json_array())).get_result::<bool>(conn)?;
+    /// assert_eq!(res, false);
+    /// let res = diesel::select(("\"abc\"".into_sql::<Text>().is_not_json_array())).get_result::<bool>(conn)?;
+    /// assert_eq!(res, true);
+    /// #
+    /// #     Ok(())
+    /// # }
+    /// ```
+    #[allow(clippy::wrong_self_convention)] // This is named after the sql operator
+    fn is_not_json_array(self) -> dsl::IsNotJsonArray<Self> {
+        IsNotJsonArray::new(self)
+    }
+
+    /// Creates a PostgreSQL `IS JSON SCALAR` expression.
+    /// Requires PostgreSQL>=16
+    ///
+    /// This operator returns true whether an object is a valid JSON SCALAR
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// # include!("../../doctest_setup.rs");
+    /// #
+    /// # fn main() {
+    /// #     run_test().unwrap();
+    /// # }
+    /// #
+    /// # fn run_test() -> QueryResult<()> {
+    /// #     use std::collections::Bound;
+    /// #     use diesel::sql_types::{Text, Nullable};
+    /// #     use diesel::dsl::sql;
+    /// #
+    /// #     let conn = &mut establish_connection();
+    /// #
+    ///
+    /// let res = diesel::select(("123".into_sql::<Text>().is_json_scalar())).get_result::<bool>(conn)?;
+    /// assert_eq!(res, true);
+    /// let res = diesel::select(("abc".into_sql::<Text>().is_json_scalar())).get_result::<bool>(conn)?;
+    /// assert_eq!(res, false);
+    /// let res = diesel::select(("{\"products\": [1,2,3]}".into_sql::<Text>().is_json_scalar())).get_result::<bool>(conn)?;
+    /// assert_eq!(res, false);
+    /// let res = diesel::select(("[1,2,3]".into_sql::<Text>().is_json_scalar())).get_result::<bool>(conn)?;
+    /// assert_eq!(res, false);
+    /// let res = diesel::select(("\"abc\"".into_sql::<Text>().is_json_scalar())).get_result::<bool>(conn)?;
+    /// assert_eq!(res, true);
+    /// let res = diesel::select(sql::<Nullable<Text>>("NULL").is_json_scalar().nullable()).get_result::<Option<bool>>(conn)?;
```

Comment:
I'm sorry to bring this up now, but this is an issue. In this case `is_json_scalar()` should evaluate to `Nullable<Bool>` instead. It's actually a short coming of the `postfix_operator!()` macro that I forgot about. The `diesel::infix_operator!()` already handles this internally: https://github.com/diesel-rs/diesel/blob/2090344bf4b7259a262ea80311ba2c33b7e66c12/diesel/src/expression/operators.rs#L296-L330

Something like this is also needed for the `postfix_operator!()` macro, although:

* We need to make sure that's not the new default, as that would be a breaking chaneg
* It's likely simpler as we just have one argument instead of two.

We likely just need to introduce an explict opt-in branch to this macro that sets the return type to 

```
$crate::sql_types::is_nullable::MaybeNullable<<T as Expression>::SqlType>
```
after that this test should just compile as

```
diesel::select(sql::<Nullable<Text>>("NULL").is_json_scalar()).get_result::<Option<bool>>(conn)?;
```

Reply:
I am not an expert when it comes to macros but I am thinking about sth like `postfix_operator!(IsJsonScalar, " IS JSON SCALAR", ConditionalNullability Bool, backend: Pg);` to make it backward compatible. @weiznich , what do you think ?

Reply:
That would be a fine solution :+1:

Reply:
@weiznich , I added ConditionalNullability support to postfix_operator. Please let me know if it makes sense.